src/QMCDrivers/WalkerControlBase.cpp

```diff
@@ -443,6 +442,105 @@ int WalkerControlBase::sortWalkers(MCWalkerConfiguration& W)
   return NumWalkers;
 }
 
+int WalkerControlBase::applyNmaxNmin()
+{
+  int current_population = std::accumulate(NumPerNode.begin(), NumPerNode.end(), 0);
+
+  // limit Nmax
+  int current_max = (current_population + NumContexts - 1) / NumContexts;
+  if(current_max>Nmax)
+  {
+    app_warning() << "Exceeding Max Walkers per MPI rank : "
+                  << Nmax << ". Ceiling is applied" << std::endl;
+    int nsub = current_population - Nmax * NumContexts;
+    for(int inode=0; inode<NumContexts; inode++)
+      if(NumPerNode[inode]>Nmax)
+      {
+        int n_remove = std::min(nsub, NumPerNode[inode]-Nmax);
+        NumPerNode[inode] -= n_remove;
+        nsub -= n_remove;
+
+        if(inode==MyContext)
+        {
+          for(int iw=0; iw<ncopy_w.size(); iw++)
+          {
+            int n_remove_walker = std::min(ncopy_w[iw], n_remove);
+            ncopy_w[iw] -= n_remove_walker;
+            n_remove -= n_remove_walker;
+            if(n_remove==0) break;
+          }
+
+          if(n_remove>0)
+          {
+            app_warning() << "Removing copies of good walkers is not enough. "
+                          << "Removing good walkers." << std::endl;
+            do
+            {
+              bad_w.push_back(good_w.back());
+              good_w.pop_back();
+              ncopy_w.pop_back();
+              --n_remove;
+            } while(n_remove>0 && !good_w.empty());
+          }
+
+          if(n_remove)
+            APP_ABORT("WalkerControlBase::applyNmax not able to remove sufficient walkers on a node!");
+          if(std::accumulate(ncopy_w.begin(), ncopy_w.end(), ncopy_w.size())!=NumPerNode[inode])
+            APP_ABORT("WalkerControlBase::applyNmax walker removal mismatch!");
+        }
+
+        if(nsub==0) break;
+      }
+
+    if(nsub) APP_ABORT("WalkerControlBase::applyNmax not able to remove sufficient walkers overall!");
+  }
+
+  // at least one walker after load-balancing
+  int current_min = current_population / NumContexts;
+  if(current_min==0)
```

Comment:
Is there a reason to care if we have one walker per rank? Physically, the only true error condition is whether there are any walkers left. The parallelization is not relevant here. My preference would be to only check the total current population. For example running on 1 million ranks with around 1 million walkers should not cause the code to stop if we have 999999 walkers. I could also choose to run on 1 million ranks with an initial 800000 walkers, which might be reasonable if I know there will be huge population fluctuation. (Note that I think the original logic in the code was wrong, not your patch.)

Reply:
I agree with your example. However, it can not be done without other changes in the code. In the branching part, we only have the capability to copy a walker not initialize it from scratch. To make a copy of a walker, you need at least 1 good or bad walker. The current code also clear all the bad walkers at the end of the load-balancing. Once you have zero walker, that node is in trouble. This is the limitation of the current code.

Reply:
OK, so we can merge after others have had a chance to comment.